packages/remix-server-runtime/dev.ts

```diff
@@ -25,3 +25,22 @@ export async function broadcastDevReady(build: ServerBuild, origin?: string) {
 export function logDevReady(build: ServerBuild) {
   console.log(`[REMIX DEV] ${build.assets.version} ready`);
 }
+
+type DevServerHooks = {
+  getCriticalCss: (
+    build: ServerBuild,
+    pathname: string
+  ) => Promise<string | undefined>;
+};
+
+const globalDevServerHooksKey = "__remix_devServerHooks";
+
+export function setDevServerHooks(devServerHooks: DevServerHooks) {
+  // @ts-expect-error
+  globalThis[globalDevServerHooksKey] = devServerHooks;
+}
+
+export function getDevServerHooks(): DevServerHooks | undefined {
+  // @ts-expect-error
+  return globalThis[globalDevServerHooksKey];
+}
```

Comment:
Nice! I didn't think about exposing get/set API.
If I look at this simplicity, then I wonder if it would also work with just using local `let` variable like:

```tsx
let _devServerHooks: DevServerHooks | undefined;

export function setDevServerHooks(devServerHooks: DevServerHooks) {
  _devServerHooks = devServerHooks
}

export function getDevServerHooks(): DevServerHooks | undefined {
  return _devServerHooks;
}
```

But this might have issues when dual package hazard like scenario (user-code is esm, but remix library is commonjs)?

---

I was just looking at the sveltekit for how they deal with `ssrFixStacktrace` and it looks like they have a clever way to pass "vite server thing" to "runtime":

https://github.com/sveltejs/kit/blob/b6c7e8b574b28b83407c068757a456d511891027/packages/kit/src/exports/vite/dev/index.js#L464-L467

```tsx
				const { set_fix_stack_trace } = await vite.ssrLoadModule(
					`${runtime_base}/shared-server.js`
				);
				set_fix_stack_trace(fix_stack_trace);
```

https://github.com/sveltejs/kit/blob/b6c7e8b574b28b83407c068757a456d511891027/packages/kit/src/runtime/shared-server.js#L20-L24

```tsx
export let fix_stack_trace = (error) => error?.stack;

export function set_fix_stack_trace(value) {
	fix_stack_trace = value;
}
```

I'm still digesting what's happening with this, but it might be possible to avoid `globalThis`, so I'll investigate further and I can follow up with a separate PR if I find a better way.